ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Peras/Voting/Rules.hs

```diff
@@ -0,0 +1,537 @@
+  -- ^ The most recent certificate seen by the voter
+  , latestCertOnChain :: !(WithOrigin (LatestCertOnChainView cert))
+  -- ^ The most recent certificate present in our preferred chain
+  }
+  deriving Show
+
+-- | Construct a 'PerasVotingView'.
+--
+-- NOTE: this assumes that the client code computes all the needed inputs
+-- within the same STM transaction, or the results may be inconsistent.
+mkPerasVotingView ::
+  ( cert ~ WithArrivalTime (ValidatedPerasCert blk)
+  , GetHeader blk
+  ) =>
+  -- | Peras protocol parameters
+  PerasParams ->
+  -- | Current Peras round number
+  PerasRoundNo ->
+  -- | Most recent certificate seen by the voter
+  WithOrigin cert ->
+  -- | Most recent certificate included in some block in our preferred chain
+  WithOrigin cert ->
+  -- | Prefix of the preferred chain leading to the candidate block
+  AnchoredFragment (Header blk) ->
+  -- | Constructed voting view
+  PerasQry xs (PerasVotingView cert)
+mkPerasVotingView
+  perasParams
+  currRoundNo
+  latestCertSeen
+  latestCertOnChain
+  chainAtCandidateBlock = do
+    latestCertSeenView <- traverse mkLatestCertSeenView latestCertSeen
+    latestCertOnChainView <- traverse mkLatestCertOnChainView latestCertOnChain
+    pure $
+      PerasVotingView
+        { perasParams = perasParams
+        , currRoundNo = currRoundNo
+        , latestCertSeen = latestCertSeenView
+        , latestCertOnChain = latestCertOnChainView
+        }
+   where
+    mkLatestCertSeenView cert = do
+      roundStartSlot <- perasRoundStart (getPerasCertRound cert)
+      arrivalSlot <- perasCertArrivalSlot cert
+      pure $
+        LatestCertSeenView
+          { lcsCert = cert
+          , lcsArrivalSlot = arrivalSlot
+          , lcsRoundStartSlot = roundStartSlot
+          , lcsCandidateBlockExtendsCert = candidateBlockExtendsCert cert
+          }
+
+    mkLatestCertOnChainView cert =
+      pure $
+        LatestCertOnChainView
+          { lcocCert = cert
+          }
+
+    -- Does the candidate block extend the one boosted by a certificate?
+    --
+    -- This can be trivially tested by checking whether the certificate is
+    -- within the bounds of the chain prefix leading to the candidate block.
+    -- Conversely, the case of a certificate pointing to a block that's too new
+    -- to be voted for is also covered by this logic, as it won't be part of the
+    -- 'chainAtCandidateBlock' fragment.
+    --
+    -- NOTE: in the case of an extremely old certificate boosting a block beyond
+    -- the immutable prefix, this could incorrectly return false even if the
+    -- voting candidate technically extends the certificate's boosted block.
+    -- This case /should not/ affect the voting rules: this value is only used
+    -- for the "happy path", and the incorrect value can only happen during a
+    -- cooldown period where the "happy path" rules are not relevant.
+    --
+    -- Nevertheless, we should fix this by enhancing our view of the latest
```

Comment:
I would prefer if `mkPerasVotingView` could take a boolean (or use a newtype around `lastCertSeen`) to indicate whether or not it is pointing to a block beyond the immutable horizon.

This way we put the responsiblity for a fix/mitigation on the caller, not the internal implementation.